crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -683,4 +689,30 @@ mod test {
         stage.run(&mut world);
         assert!(*world.get_resource::<bool>().unwrap(), "after test");
     }
+
+    #[test]
+    fn repeated_on_in_stack_triggers() {
+        let mut world = World::default();
+
+        world.insert_resource(0u32);
+        world.insert_resource(State::new(MyState::S1));
+
+        let mut stage = SystemStage::parallel();
+
+        stage.add_system_set(State::<MyState>::get_driver());
+
+        stage.add_system_set(
+            SystemSet::on_in_stack_update(MyState::S1).with_system(
+                (|mut r: ResMut<u32>, mut s: ResMut<State<MyState>>| {
+                    *r += 1;
+                    s.push(MyState::S2).ok();
+                })
+                .system(),
+            ),
+        );
```

Comment:
We could remove `.system()` to make it look nicer because of changes in main :)
```suggestion
        stage.add_system_set(SystemSet::on_in_stack_update(MyState::S1).with_system(
            |mut counter: ResMut<u32>, mut state: ResMut<State<MyState>>| {
                *counter += 1;
                state.push(MyState::S2).ok();
            },
        ));
```